lib/cloud/gcp/vm.go

```diff
@@ -0,0 +1,545 @@
+				newKeys = append(newKeys, key)
+			}
+		}
+		item.Value = googleapi.String(strings.Join(newKeys, "\n"))
+		return
+	}
+}
+
+// RemoveSSHKey removes an SSH key from a GCP VM's metadata.
+func (clt *instancesClient) RemoveSSHKey(ctx context.Context, req *SSHKeyRequest) error {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	removeSSHKey(req.Instance.metadata)
+	op, err := clt.InstanceClient.SetMetadata(ctx, &computepb.SetMetadataInstanceRequest{
+		Instance:         req.Instance.Name,
+		MetadataResource: req.Instance.metadata,
+		Project:          req.Instance.ProjectID,
+		Zone:             req.Instance.Zone,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := op.Wait(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// RunCommandRequest contains parameters for running a command on an instance.
+type RunCommandRequest struct {
+	// Client is the instance client to use.
+	Client InstancesClient
+	// InstanceRequest is the set of parameters identifying the instance.
+	InstanceRequest
+	// Script is the script to execute.
+	Script string
+	// SSHPort is the ssh server port to connect to. Defaults to 22.
+	SSHPort     string
+	dialContext func(ctx context.Context, network, addr string) (net.Conn, error)
+}
+
+func (req *RunCommandRequest) CheckAndSetDefaults() error {
+	if err := req.InstanceRequest.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if req.Script == "" {
+		return trace.BadParameter("script must be set")
+	}
+	if req.SSHPort == "" {
+		req.SSHPort = "22"
+	}
+	if req.dialContext == nil {
+		dialer := net.Dialer{}
+		req.dialContext = dialer.DialContext
+	}
+	return nil
+}
+
+func generateKeyPair() (ssh.Signer, ssh.PublicKey, error) {
+	rawPriv, rawPub, err := native.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	signer, err := ssh.ParsePrivateKey(rawPriv)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	publicKey, _, _, _, err := ssh.ParseAuthorizedKey(rawPub)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	return signer, publicKey, nil
+}
+
+// RunCommand runs a command on an instance.
+func RunCommand(ctx context.Context, req *RunCommandRequest) error {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Generate keys and add them to the instance.
+	signer, publicKey, err := generateKeyPair()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	instance, err := req.Client.GetInstance(ctx, &req.InstanceRequest)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(instance.hostKeys) == 0 {
+		return trace.NotFound(`Instances %v is missing host keys. Did you enable guest attributes on the instance?
+https://cloud.google.com/solutions/connecting-securely#storing_host_keys_by_enabling_guest_attributes`, req.Name)
+	}
+	if instance.ipAddress == "" {
+		return trace.NotFound("Instance %v is missing a public ip address.", req.Name)
```

Comment:
Why is this a requirement? If you deploy a discovery service in a VPC that can connect to your internal VMs, then nodes don't need to have public IPs, right?

I think this is a configuration we need to support.